### PR TITLE
perf(socketio): optimize Packet: TryFrom<&Bytes>

### PR DESCRIPTION
**Pull request by MaxOhn**
Rewrite of the `Packet: TryFrom<&Bytes>` implementation.

I ran a benchmark on randomly generated packets and saw a performance improvement of up to 15x 😮
Can push the benchmark to a public repository if you want to verify.

This also gets rid of the direct dependency on `regex`.

Haven't touched the test coverage since it seems decent already. If you want I can try figuring out and adding some more test cases.

### CODE REVIEW

File: socketio/src/packet.rs [hunk 1]
```diff
@@ -154,102 +153,61 @@ impl TryFrom<&Bytes> for Packet {
...
-        let mut utf8_iter = payload_utf8.chars().peekable();
-        let mut next_utf8;
-        let mut char_buf: Vec<char> = vec![];
+        let mut payload = str_from_utf8(&payload).map_err(Error::InvalidUtf8)?;
+        let mut packet = Packet::default();
 
         // packet_type
-        packet.packet_type =
-            PacketId::try_from(utf8_iter.next().ok_or(Error::IncompletePacket())?)?;
+        let c = payload.chars().next().ok_or(Error::IncompletePacket())?;
```
Comment (1c3t3a) [inline socketio/src/packet.rs:160]: nit: Can you please pick a more descriptive name than 'c'.

**Review by 1c3t3a — APPROVED** [review]
LGTM! Great, really happy to hear that about 15x perf improvement, very well appreciated :) 

Just out of curiousity, what part of the code do you think makes the difference?

**Comment by 1c3t3a** [comment]
Can you please drop the merge commit and instead rebase the other changes? :)

**Comment by MaxOhn** [comment]
> LGTM! Great, really happy to hear that about 15x perf improvement, very well appreciated :)
> 
> Just out of curiousity, what part of the code do you think makes the difference?

It's a combination of multiple things I would say.
- The `json_data` was not only validated but also used to create the string. `serde_json::Value::to_string` is pretty costly and wasn't even necessary because the json string it was built with could have been used instead.
- `.to_vec()`, `.to_string()`, `.collect()`, `.to_owned()`, and even `str::replace` each come with an allocation, never a good sign.
- Building a `Regex` at runtime is pretty expensive. As luck would have it, the most recent rust version 1.70 introduced [`OnceLock`](https://doc.rust-lang.org/std/sync/struct.OnceLock.html) which would have allowed to compile it once and re-use it but the better alternative of course is to not use regex at all in the first place.
- I replaced a bunch of manual loops and their branching. Iterator methods are often times faster since rust is so fine-tuned for those. And using std lib methods wherever applicable in general is likely a good idea, they consist of performant magic. 😄

**Comment by MaxOhn** [comment]
Oh apologies, didn't mean to close / merge this just yet with the force push after rebasing 😅

**Pull request closed without merging**